lib/collection/src/collection/query.rs

```diff
@@ -0,0 +1,122 @@
+                .and_then(move |mut records| async move {
+                    if shard_key.is_none() {
+                        return Ok(records);
+                    }
+                    for batch in &mut records {
+                        for point in batch {
+                            point.shard_key.clone_from(&shard_key);
+                        }
+                    }
+                    Ok(records)
+                })
+        });
+        future::try_join_all(all_searches).await
+    }
+
+    /// This is a remote shard implementation to querying the shards. Only used for the internal service
+    pub async fn query_internal(
+        &self,
+        request: ShardQueryRequest,
+        read_consistency: Option<ReadConsistency>,
+        shard_selection: &ShardSelectorInternal,
+    ) -> CollectionResult<ShardQueryResponse> {
+        let request = Arc::new(request);
+
+        let mut all_shards_results = self
+            .query_shards_concurrently(Arc::clone(&request), read_consistency, shard_selection)
+            .await?;
+
+        let (num_intermediate_responses, intermediate_infos) =
+            if let Some(ScoringQuery::Fusion(Fusion::Rrf)) = request.query {
+                (
+                    request.prefetches.len(),
+                    Either::Left(
+                        request
+                            .prefetches
+                            .iter()
+                            .map(|prefetch| IntermediateQueryInfo {
+                                scoring_query: prefetch.query.as_ref(),
+                                take: prefetch.limit,
+                            }),
+                    ),
+                )
+            } else {
+                (
+                    1,
+                    Either::Right(iter::once(IntermediateQueryInfo {
+                        scoring_query: request.query.as_ref(),
+                        take: request.offset + request.limit,
+                    })),
+                )
+            };
+
+        let collection_params = self.collection_config.read().await.params.clone();
+
+        let mut result = ShardQueryResponse::with_capacity(num_intermediate_responses);
+
+        for (idx, intermediate_info) in intermediate_infos.enumerate() {
+            let intermediate_result_for_shard = all_shards_results
+                .iter_mut()
+                .map(|intermediates| mem::take(&mut intermediates[idx]));
+
+            let order = collection_params.scoring_order(intermediate_info.scoring_query)?;
+
+            let intermediate_result = match order {
+                Order::LargeBetter => Either::Left(
+                    intermediate_result_for_shard
+                        .kmerge_by(|a, b| ScoredPointTies(a) > ScoredPointTies(b)),
+                ),
+                Order::SmallBetter => Either::Right(
+                    intermediate_result_for_shard
+                        .kmerge_by(|a, b| ScoredPointTies(a) < ScoredPointTies(b)),
```

Comment:
To discuss: we need this to merge points which have equal scoring, like those coming from scrolls. Should we just add this comparison by ID to the `Ord` impl of `ScoredPoint`?

Reply:
Not sure I'd like that. Comparing in that way might not always make sense when you have a scored point, or does it?

The construction of `ScoredPointTies(a)` is cheap, so I wouldn't worry about it from a performance perspective.

Do you have a specific reason on why you might prefer implementing Ord on the scored point directly for this?

Reply:
> The construction of ScoredPointTies(a) is cheap, so I wouldn't worry about it from a performance perspective.

My same take on it.

I was only suggesting if it should become a part of the behavior of ScoredPoint, but we might not want that.